### PR TITLE
Add shear parallel to y-axis

### PR DESCRIPTION
**Pull request by ptrblck**
This PR allows an additional shear parallel to the y-axis as discussed in [this topic](https://discuss.pytorch.org/t/how-to-use-torchvision-transform-shearing-for-x-and-y-direction/49371/3?u=ptrblck).

The functional API accepts `shear` now also as a `tuple` or a `list`:
```python
TF.affine(
    img,
    angle=0,
    translate=[0, 0],
    scale=1.,
    shear=[0,10]
)
```
The first values will be used for a shear parallel to the x axis, while the second one will be used for a shear parallel to the y axis.

BC is maintained and the argument can still be a single value. In that case, only the x-axis shear will be applied.

The `transforms.RandomAffine` module now accepts a `tuple` or `list` of up to 4 values, where the first two values will be used to randomly sample a degree for the x-axis shear and the latter two values for the y-axis shear.
BC is maintained similar to the functional API.

Sample code:
```python
x = torch.zeros(3, 100, 100)
x[:, 25:75, 25:75] = 1.
img = TF.to_pil_image(x)

rand_shear = transforms.RandomAffine(
    degrees=0,
    shear=10)
rand_shear(img)

rand_shear = transforms.RandomAffine(
    degrees=0,
    shear=[10, 20])
rand_shear(img)


rand_shear = transforms.RandomAffine(
    degrees=0,
    shear=[10., 20., 40, 50])
rand_shear(img)

rand_shear = transforms.RandomAffine(
    degrees=0,
    shear=[0, 0, -10, 20])
rand_shear(img)
```

The added tests have all passed, but let me know, if we need additional tests.

CC @fmassa

### CODE REVIEW

File: torchvision/transforms/functional.py [hunk 1]
```diff
@@ -728,14 +728,20 @@ def _get_inverse_affine_matrix(center, angle, translate, scale, shear):
     # Thus, the inverse is M^-1 = C * RSS^-1 * C^-1 * T^-1
 
     angle = math.radians(angle)
-    shear = math.radians(shear)
+    print(shear)
```
Comment (fmassa) [inline torchvision/transforms/functional.py]: Can you remove this print?

File: torchvision/transforms/functional.py [hunk 2]
```diff
@@ -728,14 +728,20 @@ def _get_inverse_affine_matrix(center, angle, translate, scale, shear):
...
+    if isinstance(shear, (tuple, list)) and len(shear) == 2:
```
Comment (fmassa) [inline torchvision/transforms/functional.py]: What if the length of `shear` is `> 2`, or `== 1`? This will go into the next `else` block, and will fail.

Maybe we could add an assert before this part of the code to assert those constraints?

File: torchvision/transforms/functional.py (hunk 1 above)
Comment (ptrblck) [reply torchvision/transforms/functional.py]: Sure! Haven't noticed this debug statement.

**Comment by vfdev-5** [comment]
@ptrblck @fmassa newer API seems ok for me.

@ptrblck nit, maybe we would to update commented description line 725
```
    #       RSS(a, scale, shear) = [ cos(a)*scale    -sin(a + shear)*scale     0]
    #                              [ sin(a)*scale    cos(a + shear)*scale     0]
    #                              [     0                  0          1]
```
and make sure that new `d`and `matrix` correspond to the newer API
```python
    d = math.cos(angle + shear[0]) * math.cos(angle + shear[1]) + \
        math.sin(angle + shear[0]) * math.sin(angle + shear[1])
    matrix = [
        math.cos(angle + shear[0]), math.sin(angle + shear[0]), 0,
        -math.sin(angle + shear[1]), math.cos(angle + shear[1]), 0
    ]
```

File: torchvision/transforms/functional.py (hunk 2 above)
Comment (ptrblck) [reply torchvision/transforms/functional.py]: I've added an `else` statement to handle these errors and raise an exception, as I wanted to avoid code duplication (checking the length twice: in the assert and again to create the shear list).
What do you think?

Comment (fmassa) [reply torchvision/transforms/functional.py]: This looks ok to me!

**Comment by jacobhinkle** [comment]
Hi everyone,

I just noticed this PR and would like to point out that I believe this change corresponds to replacing a shear that looks like `[1 sin(u); 0 1]` with `[1 sin(u); sin(v) 1]`. The first matrix is a shear in X as was the previous behavior, and clearly when `u=0` the second is a shear in Y. But simultaneous "shearing" in this way is not a product of shears and in fact is bound to induce some shrinking since `det([1 sin(u); sin(v) 1])=1-sin(u)sin(v)`. For example when `u=v=pi/2` this is a singular matrix, while the `v=0` case corresponding to `[1 1; 0 1]` is in some sense the "stereotypical" shear.

Taking this further, insert `u=v=\pi` into the formula (with no rotation or scale) to see that this is an inversion through the origin: `RSS = [-1 0; 0 -1]`.

I am unfamiliar with tensorflow, so I don't know how they handle this. It may not be much of an issue since typically some rotation and scaling is desired if shear is being considered as well, but it is probably a good idea to have a handle on the expected determinant of RSS since unless I've made an error it is no longer the expected scale.

**Comment by fmassa** [comment]
@jacobhinkle thanks for your feedback. Do you have any recommendations on what should be done here?

**Comment by ptrblck** [comment]
Hi @jacobhinkle,

wouldn't the transformation matrix for `u=v=pi/2` be:
```python
[[0, 1],
 [-1, 0]]
```
with `det(A)=1`, which is thus not singular?
Please correct me, if I'm missing something.

I assumed the shrinking is wanted behavior.
If we ignore this PR and think about the x-axis shear with an angle of pi/2:
since we collapse the image into a line (`det(A)=0`), I'm not sure, how we could keep the determinant constant. Should these values throw an error instead or should the determinant be only "fixed" when we have a shear parallel to both axes?
If so, how would we handle the case for `shear = [45, -45+eps]`, which will also lose a degree of freedom.

Maybe I'm seeing it too naively, so let me know, what the desired behavior would be. :)

**Comment by jacobhinkle** [comment]
Sorry, you are right that `u=v=pi/2` leads to the rotation you stated. But `u=pi/2, v=-pi/2` for example would give `[0 1; 1 0]` I believe. I think a simple way to handle it is to choose one element on the diagonal and add `u*v` to that diagonal, resulting in `[1 u; v 1+u*v]`. That is equivalent to choosing for example to apply and X shear followed by a Y shear, as in the example linked in the forum post. I think this leads to the following RSS matrix, but would welcome some double-checking:
```
ct, st = cos(theta), sin(theta)
cu, su = cos(u), sin(u)
cv, sv = cos(v), sin(v)
RSS = [[ ct - st*sv, ct*su - st*(1+su*sv)],
       [ st + ct*sv, st*su + ct*(1+su*sv)]]
```
There are an additional two trig functions there, and I'm not sure if it simplifies. This was derived as the product
```
RSS = [[ct, -st], [st, ct]] @ [[1, 0],[sv, 1]] @ [[1, su],[0, 1]]
```

Regarding an X-axis shear of pi/2, doesn't that correspond to the matrix `[1 1; 0 1]` using the `sin(theta)` parametrization. That is what I thought was being used in the original formulation but I may be way off-base. If pi/2 is meant to be the actual shear angle I believe we should insert cotangent(pi/2-theta) instead of sin(theta), in which case that entry is infinite when theta is pi/2. Using `cot(pi/2-theta)` parametrization, it seems the original formula is missing a `/sin(pi/2-theta)` for the `cos(a+theta)` and `sin(a+theta)` terms. I will have some more time this afternoon to think this through.
